README.md
# slug_ts

Slug generator for url or human friendly ids written as a tutorial for TypeScript clean and tested code

## 👨 Created by Alberto Basalo

[@albertobasalo](https://twitter.com/albertobasalo)

### There is a [JavaScript version](https://github.com/LabsAdemy/slug_js) of this project


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [1.0.2](https://github.com/AtomicBuilders/slug_ts/compare/v1.0.1...v1.0.2) (2021-06-21)

### 1.0.1 (2021-06-21)

## [1.6.0](https://github.com/AtomicBuilders/quark/compare/v1.5.2...v1.6.0) (2021-06-21)


### Features

* config and scripts ([b242838](https://github.com/AtomicBuilders/quark/commit/b2428386a95f3ffb8d5ad5c9af6679035f66c8e4))

### [1.5.2](https://github.com/AtomicBuilders/quark/compare/v1.5.1...v1.5.2) (2021-06-09)


### Bug Fixes

* release comand to main branch ([04f2c7a](https://github.com/AtomicBuilders/quark/commit/04f2c7a595b0233955002a24674453073e39597f))

### [1.5.1](https://github.com/AtomicBuilders/quark/compare/v1.5.0...v1.5.1) (2021-06-09)

## [1.5.0](https://github.com/AtomicBuilders/quark/compare/v1.4.0...v1.5.0) (2021-01-08)


### Features

* TypeScript version ([41a316f](https://github.com/AtomicBuilders/quark/commit/41a316f011fe96173e15bd1ada3f42bf91ab6b49)), closes [#18](https://github.com/AtomicBuilders/quark/issues/18)
* workflows documentation Closes [#19](https://github.com/AtomicBuilders/quark/issues/19) ([1565fc7](https://github.com/AtomicBuilders/quark/commit/1565fc7a4ab78415433cdaff5ca96d9d4d02dd4b))

## [1.4.0](https://github.com/AtomicBuilders/quark/compare/v1.3.1...v1.4.0) (2021-01-07)


### Features

* TypeScript version ([6fa9985](https://github.com/AtomicBuilders/quark/commit/6fa998508f15969f3773436abfa5b3d1c203719c)), closes [#18](https://github.com/AtomicBuilders/quark/issues/18)

### [1.3.1](https://github.com/AtomicBuilders/quark/compare/v1.3.0...v1.3.1) (2021-01-05)


### Bug Fixes

* readme ([fc4c38d](https://github.com/AtomicBuilders/quark/commit/fc4c38d1ab1a0c8592b4498f75c73e4fd757efd2))

## [1.3.0](https://github.com/AtomicBuilders/quark/compare/v1.2.1...v1.3.0) (2021-01-05)


### Features

* settings and extensions ([e000f1a](https://github.com/AtomicBuilders/quark/commit/e000f1a94156989083f822870f0ff2c76169c5c7))


### Bug Fixes

* config ([490ca90](https://github.com/AtomicBuilders/quark/commit/490ca90b4a44460d39a3944dce8fb8c95249044f))

### [1.2.1](https://github.com/AtomicBuilders/quark/compare/v1.2.0...v1.2.1) (2021-01-05)

## [1.2.0](https://github.com/AtomicBuilders/quark/compare/v1.1.6...v1.2.0) (2021-01-04)


### Features

* eslint ([34ef7ca](https://github.com/AtomicBuilders/quark/commit/34ef7ca9fb76474979bf747398f03a39552549c3))

### [1.1.6](https://github.com/AtomicBuilders/quark/compare/v1.1.5...v1.1.6) (2021-01-04)

### [1.1.5](https://github.com/AtomicBuilders/quark/compare/v1.1.4...v1.1.5) (2020-12-30)

### [1.1.4](https://github.com/AtomicBuilders/quark/compare/v1.1.3...v1.1.4) (2020-09-17)


### Bug Fixes

* script final coma ([d8685a5](https://github.com/AtomicBuilders/quark/commit/d8685a57c3e2f255b2eb6e2c638f1ce0b5bf9c15))

### [1.1.3](https://github.com/AtomicBuilders/quark/compare/v1.1.2...v1.1.3) (2020-09-17)

### [1.1.2](https://github.com/AtomicBuilders/quark/compare/v1.1.1...v1.1.2) (2020-09-10)

### [1.1.1](https://github.com/AtomicBuilders/quark/compare/v1.1.0...v1.1.1) (2019-06-20)



## [1.1.0](https://github.com/AtomicBuilders/quark/compare/v1.0.2...v1.1.0) (2019-06-20)


### Features

* **GitHub:** scripts to manage issue labels ([816a937](https://github.com/AtomicBuilders/quark/commit/816a937))



### [1.0.2](https://github.com/AtomicBuilders/quark/compare/v1.0.1...v1.0.2) (2019-06-20)



# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.


src/lib.spec.ts
import { getSlug, remove, replace, slugOptions } from "./lib";
describe("GIVEN: the getSlug function used without options and a string source", () => {
  it("WHEN has spaces THEN should be replaced with a hyphen", () => {
    const input = "3 concepts you need to get started";
    const actual = getSlug(input);
    const expected = "3-concepts-you-need-to-get-started";
    expect(actual).toStrictEqual(expected);
  });
  it("WHEN has capital letters THEN should be on lowercase", () => {
    const input = "Elon Musk considers move to Mars";
    const actual = getSlug(input);
    const expected = "elon-musk-considers-move-to-mars";
    expect(actual).toStrictEqual(expected);
  });
  it("WHEN has € THEN should be on replaced with euros", () => {
    const input = "Fintech startups raised 34B€ in 2019";
    const actual = getSlug(input);
    const expected = "fintech-startups-raised-34beuros-in-2019";
    expect(actual).toStrictEqual(expected);
  });
  it("WHEN has an apostrophe THEN should be removed", () => {
    const input = "Shopify joins Facebook’s cryptocurrency Libra Association";
    const actual = getSlug(input);
    const expected = "shopify-joins-facebooks-cryptocurrency-libra-association";
    expect(actual).toStrictEqual(expected);
  });
  it("WHEN has question marks THEN should be removed", () => {
    const input = "What is a slug and how to optimize it?";
    const actual = getSlug(input);
    const expected = "what-is-a-slug-and-how-to-optimize-it";
    expect(actual).toStrictEqual(expected);
  });
  it("WHEN has commas THEN should be removed", () => {
    const input = "The COVID-19, coronavirus desease, is being defeated";
    const actual = getSlug(input);
    const expected = "the-covid-19-coronavirus-desease-is-being-defeated";
    expect(actual).toStrictEqual(expected);
  });
  it("WHEN has colons THEN should be removed", () => {
    const input =
      "Design Patterns: Top 5 Techniques for Implementing Fault Tolerance in Distributed Systems";
    const actual = getSlug(input);
    const expected =
      "design-patterns-top-5-techniques-for-implementing-fault-tolerance-in-distributed-systems";
    expect(actual).toStrictEqual(expected);
  });
  it("WHEN is too long THEN should be truncated to 100 chars", () => {
    const input =
      "Running our test suite with Mocha took 12+ minutes. In CI with our beefy build machines we’re now able to run the entire Jest suite in 4 minutes 30 seconds";
    const actual = getSlug(input);
    const expected =
      "running-our-test-suite-with-mocha-took-12-minutes-in-ci-with-our-beefy-build-machines-were-now-able-";
    expect(actual).toStrictEqual(expected);
  });
});

describe("GIVEN: the getSlug function used with options", () => {
  it("WHEN we want a timestamp THEN should be added to the end", () => {
    const inputSource =
      "When should I migrate my Frontend to an architecture based on microfrontends?";
    const inputOptions: Partial<slugOptions> = { timestamp: true };
    const actual = getSlug(inputSource, inputOptions);
    const expected =
      "when-should-i-migrate-my-frontend-to-an-architecture-based-on-microfrontends_";
    expect(actual.startsWith(expected)).toBeTruthy();
    expect(actual.length - expected.length).toStrictEqual(13);
  });
  it("WHEN we want a datestamp THEN should be added to the end", () => {
    const inputSource = "Solve Module Import Aliasing for Webpack, Jest, and VSCode";
    const inputOptions: Partial<slugOptions> = { datestamp: true };
    const actual = getSlug(inputSource, inputOptions);
    const expected = "solve-module-import-aliasing-for-webpack-jest-and-vscode_";
    expect(actual.startsWith(expected)).toBeTruthy();
    expect(actual.length - expected.length).toStrictEqual(10);
  });
  it("WHEN we want several options THEN should be added to the end", () => {
    const inputSource = "What Is A Smart Contract And How Does It Work ?";
    const inputOptions: Partial<slugOptions> = {
      length: 30,
      timestamp: true,
    };
    const actual = getSlug(inputSource, inputOptions);
    const expected = "what-is-a-smart-contract-and-h_";
    expect(actual.startsWith(expected)).toBeTruthy();
    expect(actual.length - expected.length).toStrictEqual(13);
    expect(actual).toHaveLength(30 + 1 + 13);
  });
});

describe("GIVEN the replace function", () => {
  it("WHEN gets a $ THEN returns dollars", () => {
    const input = "Bitcoins drops under 50000$";
    const actual = replace(input, "$", "dollars");
    const expected = "Bitcoins drops under 50000dollars";
    expect(actual).toStrictEqual(expected);
  });
});

describe("GIVEN the remove function", () => {
  it("WHEN gets an array of targets $ THEN returns without them", () => {
    const input = "Learn JavaScript, Testing, Async/Await & more following @albertobasalo";
    const inputTargets = ["’", "!", "?", ",", ":", ".", "#", "+", "/", "&", "@"];
    const actual = remove(input, inputTargets);
    const expected = "Learn JavaScript Testing AsyncAwait  more following albertobasalo";
    expect(actual).toStrictEqual(expected);
  });
});


src/main.spec.ts
import { sayHello, sum } from "./main";
describe("the main program", () => {
  it("sayHello function exists", () => {
    expect(sayHello).toBeDefined();
  });
});

describe("the sum function", () => {
  const sut = sum;
  it("should sum numbers", () => {
    const actual = sut(1, 2);
    const expected = 3;
    expect(actual).toBe(expected);
  });
});


src/lib.ts
export function getSlug(source: string, slugOptions?: Partial<slugOptions>): string {
  const options: slugOptions = {
    length: 100,
    timestamp: false,
    datestamp: false,
    replaceFn: replace,
    removeFn: remove,
    ...slugOptions,
  };

  let cleanSource = String(source).trim().toLowerCase();
  const targets = ["’", "!", "?", ",", ":", ".", "#", "+", "/", "&", "@"];
  cleanSource = options.removeFn(cleanSource, targets);
  cleanSource = options.replaceFn(cleanSource, " ", "-");
  cleanSource = options.replaceFn(cleanSource, "$", "dollars");
  cleanSource = options.replaceFn(cleanSource, "€", "euros");

  if (cleanSource.length > options.length) {
    cleanSource = cleanSource.slice(0, options.length);
  }
  if (options.timestamp) {
    cleanSource += "_" + Date.now();
  }
  if (options.datestamp) {
    cleanSource += "_" + new Date().toISOString().slice(0, 10);
  }
  return cleanSource;
}

export type slugOptions = {
  length: number | 100;
  timestamp: boolean;
  datestamp: boolean;
  replaceFn: (source: string, target: string, change: string) => string;
  removeFn: (source: string, targets: string[]) => string;
};

export function replace(source: string, target: string, change: string): string {
  return source.split(target).join(change);
}

export function remove(source: string, targets: string[]): string {
  let result = source;
  targets.forEach(target => {
    result = replace(result, target, "");
  });
  return result;
}


src/main.ts
import { getSlug } from "./lib";

console.log(getSlug("3 concepts you need to get started"));

console.log(getSlug("Fintech startups raised 34B€ in 2019"));

console.log(getSlug("Shopify joins Facebook’s cryptocurrency Libra Association"));

console.log(getSlug("What is a slug and how to optimize it?"));

console.log(getSlug("The COVID-19, coronavirus desease, is being defeated"));

console.log(
  getSlug(
    "Design Patterns: Top 5 Techniques for Implementing Fault Tolerance in Distributed Systems"
  )
);

console.log(
  getSlug(
    "Running our test suite with Mocha took 12+ minutes. In CI with our beefy build machines we’re now able to run the entire Jest suite in 4 minutes 30 seconds"
  )
);

console.log(
  getSlug("When should I migrate my Frontend to an architecture based on microfrontends?", {
    timestamp: true,
  })
);

console.log(
  getSlug("Solve Module Import Aliasing for Webpack, Jest, and VSCode", {
    datestamp: true,
  })
);

console.log(
  getSlug("What Is A Smart Contract And How Does It Work ?", {
    length: 30,
    timestamp: true,
  })
);
